### PR TITLE
PR: file_lattice_clean

### PR DESCRIPTION
**Pull request by tuopuu**
Picked the commits from file_lattice2 into a new clean branch. This pull request adds basic functionality for H5CUDS-based lattices.
- H5CUDS now has add/get/delete/iter_lattices methods
- FileLattice implements ABCLattice in a H5CUDS-file
- unittests for new methods are included

### CODE REVIEW

File: simphony/io/tests/test_h5_cuds.py [hunk 1]
```diff
@@ -346,6 +347,40 @@ def test_mesh_rename(self):
...
+        lat2 = Lattice('test_lattice2', 'cubic', (1.0, 1.0, 1.0),
+                       (3, 3, 3), (0.0, 0.0, 0.0))
+
+        filename = os.path.join(self.temp_dir, 'test.cuds')
+        with closing(H5CUDS.open(filename, 'w')) as handle:
+            handle.add_lattice(lat)
+            handle.add_lattice(lat2)
+            for lat in handle.iter_lattices():
+                self.assertTrue(lat.name in ['test_lattice', 'test_lattice2'])
+
```
Comment (nathanfranklin) [inline simphony/io/tests/test_h5_cuds.py]: Could a test of when the lattice is renamed also be added. 

See test_mesh_rename and test_particle_container_rename

File: simphony/cuds/lattice.py [hunk 2]
```diff
@@ -26,22 +25,25 @@ def __init__(self, index, data=None):
...
 
-    Parameters
-    ----------
-    name : str
-    type : str
+    Stores references to data containers (node related data).
+
+    Attributes:
+    -----------
+    name : string
```
Comment (itziakos) [inline simphony/cuds/lattice.py]: `str` is better than `string` since `str` is the name of the string type in python

File: simphony/cuds/lattice.py [hunk 3]
```diff
@@ -26,22 +25,25 @@ def __init__(self, index, data=None):
...
+        name of the lattice
+    type : string
         Bravais lattice type (should agree with the base_vect below).
     base_vect : D x float
         defines a Bravais lattice (an alternative for primitive vectors).
-    size : tuple of D x size
+    size : tuple of D x int
```
Comment (itziakos) [inline simphony/cuds/lattice.py]: In the docstrings it is not explained what is `D`

File: simphony/io/file_lattice.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        Parameters:
+        -----------
+        index : tuple of D x int (node index coordinate)
+
+        Returns:
+        -----------
+        A reference to a LatticeNode object
+
+        """
+        n = np.ravel_multi_index(index, self._size)
```
Comment (itziakos) [inline simphony/io/file_lattice.py]: I was not sure if these methods where doing the right thing. Thanks for looking up

File: simphony/io/h5_cuds.py [hunk 5]
```diff
@@ -208,6 +262,21 @@ def delete_mesh(self, name):
...
+        """Delete lattice from file.
+
+        Parameters
+        ----------
+        name : str
+            name of lattice to delete
+        """
+        try:
+            filelat = self._handle.root.lattice._f_get_child(name)
+            filelat._f_remove(recursive=True)
```
Comment (itziakos) [inline simphony/io/h5_cuds.py]: I would move this line out of the try... except block, it is prefered to have only the code that would raise an exception in the try block. Consider the alternative
```python
try:
    filelat = self._handle.root.lattice._f_get_child(name)
except tables.NoSuchNodeError:
    raise ValueError('Lattice \'{n}\` does not exist'.format(n=name))
else:
    filelat._f_remove(recursive=True)
```

File: simphony/io/tests/test_h5_cuds.py (hunk 1 above)
Comment (tuopuu) [inline simphony/io/tests/test_h5_cuds.py]: Added a new commit that implements/fixes this functionality.

File: simphony/cuds/lattice.py (hunk 2 above)
Comment (tuopuu) [inline simphony/cuds/lattice.py]: OK, I'll use that in future.

File: simphony/io/tests/test_file_lattice.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+        shutil.rmtree(self.temp_dir)
+
+    def test_file_lattice_constructed_from_Lattice(self):
+        """ Checks that FileLattice constructed from Lattice with a
+        CustomRecord column description has correct attributes
+
+        """
+        self.assertEqual(self.filelattice.name, 'test_lattice')
+        self.assertEqual(self.filelattice.type, 'cubic')
+        self.assertItemsEqual(self.filelattice.base_vect,
```
Comment (itziakos) [inline simphony/io/tests/test_file_lattice.py]: These should probably be using assertSequenceEqual. But I think that using the `assert_array_equal` from numpy tests is probably the most appropriate method here.
you can import the method as
```python
from numpy.testing import assert_array_equal
```

File: simphony/cuds/lattice.py (hunk 3 above)
Comment (tuopuu) [inline simphony/cuds/lattice.py]: I'll add a note on the accepted values of D.

File: simphony/io/tests/test_file_lattice.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+        the nodes is returned when the function iter_nodes is called
+        without arguments
+
+        """
+        fl_nodes = self.filelattice.iter_nodes()
+
+        for M in fl_nodes:
+            N = self.lattice.get_node(M.index)
+            self.assertEqual(N.index, M.index)
+            self.assertItemsEqual(N.data, M.data)
```
Comment (itziakos) [inline simphony/io/tests/test_file_lattice.py]: This will compare only the keys in the `DataContainers`. I would suggest to use the utility methods in https://github.com/simphony/simphony-common/blob/master/simphony/testing/utils.py#L26

File: simphony/io/tests/test_file_lattice.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+        """ Checks the node iterator on a subset of nodes
+
+        """
+
+        fl_nodes = self.filelattice.iter_nodes([(0, 0, 0), (0, 1, 2)])
+
+        for M in fl_nodes:
+            N = self.lattice.get_node(M.index)
+            self.assertEqual(N.index, M.index)
+            self.assertItemsEqual(N.data, M.data)
```
Comment (itziakos) [inline simphony/io/tests/test_file_lattice.py]: See comment in https://github.com/simphony/simphony-common/pull/74/files#r25767404

File: simphony/io/tests/test_file_lattice.py [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+                        {CUBA.MATERIAL_ID: 2, CUBA.DENSITY: 10.0,
+                         CUBA.VELOCITY: (0.0, 0.0, 10.0)})
+
+        self.filelattice.update_node(N)
+
+        M = self.filelattice.get_node((3, 2, 3))
+
+        self.assertItemsEqual(N.index, M.index)
+
+        for key, value in M.data.iteritems():
```
Comment (itziakos) [inline simphony/io/tests/test_file_lattice.py]: see comment See comment in https://github.com/simphony/simphony-common/pull/74/files#r25767404

File: simphony/io/tests/test_h5_cuds.py [hunk 10]
```diff
@@ -346,6 +347,67 @@ def test_mesh_rename(self):
...
+        lat2 = Lattice('test_lattice2', 'cubic', (1.0, 1.0, 1.0),
+                       (3, 3, 3), (0.0, 0.0, 0.0))
+
+        filename = os.path.join(self.temp_dir, 'test.cuds')
+        with closing(H5CUDS.open(filename, 'w')) as handle:
+            handle.add_lattice(lat)
+            handle.add_lattice(lat2)
+            for lat in handle.iter_lattices():
+                self.assertTrue(lat.name in ['test_lattice', 'test_lattice2'])
```
Comment (itziakos) [inline simphony/io/tests/test_h5_cuds.py]: This logic will not fail iif for some reason there are two lattices with the same name or only one lattice.
I would suggest the following
```python
stored_names = [lattice.name for lattice in handle.iter_lattices()]
self.assertItemsEqual(stored_names, ['test_lattice', 'test_lattice2'])
```

File: simphony/io/tests/test_h5_cuds.py [hunk 11]
```diff
@@ -346,6 +347,67 @@ def test_mesh_rename(self):
...
+                      (2, 3, 4), (0.0, 0.0, 0.0))
+
+        filename = os.path.join(self.temp_dir, 'test.cuds')
+        with closing(H5CUDS.open(filename, 'w')) as handle:
+            handle.add_lattice(lat)
+            lat2 = handle.get_lattice('test_lattice')
+
+            self.assertEqual(lat2.name, 'test_lattice')
+
+            for N in lat2.iter_nodes():
```
Comment (itziakos) [inline simphony/io/tests/test_h5_cuds.py]: See comment in https://github.com/simphony/simphony-common/pull/74/files#r25767404

File: simphony/cuds/lattice.py (hunk 2 above)
Comment (itziakos) [inline simphony/cuds/lattice.py]: For consistancy please revert the change back to  `str`.

File: simphony/io/tests/test_file_lattice.py (hunk 6 above)
Comment (tuopuu) [inline simphony/io/tests/test_file_lattice.py]: Thanks for the tip. I'm now using assert_array_equal (without the s) for all numpy array checks.

File: simphony/io/tests/test_file_lattice.py (hunk 7 above)
Comment (tuopuu) [inline simphony/io/tests/test_file_lattice.py]: I was not aware that only keys are compared in this case. However, I couldn't make compare_data_containers work: it gives out an error "NoneType has no attribute assertEqual" on the line https://github.com/simphony/simphony-common/blob/master/simphony/testing/utils.py#L28

I don't know where does this error come from. But to move on, I'll just add manual comparison to the data with numpy.testing.assert_equal, which seems to work fine.

File: simphony/io/tests/test_h5_cuds.py (hunk 10 above)
Comment (tuopuu) [inline simphony/io/tests/test_h5_cuds.py]: You are right. I updated the code with your lines.

File: simphony/io/tests/test_file_lattice.py (hunk 7 above)
Comment (itziakos) [inline simphony/io/tests/test_file_lattice.py]: The compare function needs the test-case to be passed as a keyword argument. 
The simple way is to pass `self` as the testcase argument or register the method with the unittest machinery as in the example https://github.com/simphony/simphony-common/blob/4ad04227e7bbbeaf61f9b458eced5701b3584072/simphony/testing/abc_container_check.py#L15

File: simphony/io/h5_cuds.py [hunk 12]
```diff
@@ -177,6 +209,28 @@ def get_mesh(self, name):
...
+        file has been closed then the lattice should
+        no longer be used.
+
+        Parameters
+        ----------
+        name : str
+            name of lattice to return
+        """
+        if name in self._handle.root.lattice:
+            F = self._handle.root.lattice._v_leaves[name].attrs
```
Comment (itziakos) [inline simphony/io/h5_cuds.py]: I think this is not necessary since the ``__init__`` method of FileLattice will recover the right values.

File: simphony/io/file_lattice.py [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+    record : tables.IsDescription
+        A class that describes column types for PyTables table.
+
+    Returns:
+    ----------
+    FileLattice
+        The lattice newly added to the file or existing in the file.
+
+    """
+    def __init__(self, file, name, type, base_vect, size, origin,
```
Comment (itziakos) [inline simphony/io/file_lattice.py]: type, base_vect, size and origin should be optional since they are recovered from the hdf5 File if the lattice already exists.

File: simphony/io/h5_cuds.py (hunk 12 above)
Comment (itziakos) [inline simphony/io/h5_cuds.py]: This ofcourse assumes that https://github.com/simphony/simphony-common/pull/74/files#r26033012 has the arguments optional.

File: simphony/io/file_lattice.py [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+        """
+        if indices is None:
+            for row_number, data in enumerate(self._table):
+                index = np.unravel_index(row_number, self._size)
+                yield LatticeNode(index, data)
+        else:
+            for index in indices:
+                yield self.get_node(index)
+
+    def get_coordinate(self, index):
```
Comment (itziakos) [inline simphony/io/file_lattice.py]: It does not look that there are tests for this method

**Comment by itziakos** [comment]
@tuopuu, I had another look with only minor comment. Thanks for cleaning up the branch.

File: simphony/io/file_lattice.py (hunk 13 above)
Comment (tuopuu) [inline simphony/io/file_lattice.py]: Thank you, that's a good point. I'll change this and the related function calls in tests and h5_cuds.get_lattice.

File: simphony/io/file_lattice.py (hunk 14 above)
Comment (tuopuu) [inline simphony/io/file_lattice.py]: Thank you, I'll add it.

Comment (itziakos) [inline simphony/io/file_lattice.py]: Please make sure you add some tests for the `OrthotombicP` and `Hexagonal` lattice

**Comment by tuopuu** [comment]
@itziakos, could please take another look. I modified FileLattice constructor slightly adding None defaults to the attributes, but this creates a FutureWarning. My intention is to print out an error, if FileLattice is called with a name not existing in file, and without the optional arguments. Could you think of any other way of defining the optional arguments, or making the validity check? In my opinion, using some real values there (like,'Cubic',(1.0,1.0,1.0), etc.) would not look elegant at all. What do you think?

**Comment by nathanfranklin** [comment]
@roigcarlo , @SGGgarcia  can you all take a quick review too (time permitted).

**Comment by itziakos** [comment]
>  I modified FileLattice constructor slightly adding None defaults to the attributes, but this creates a FutureWarning, 

What is the FutureWarning, can you provide a trace back?

**Comment by SGGgarcia** [comment]
I took a quick view and looks ok :+1:

**Comment by nathanfranklin** [comment]
> I do not think that you need to check that these arguments are None. If the name is not there one should create a Lattice using default values.

It seems FileLattice is to be used when there is a lattice in a file (`get_lattice` in `H5CUDS`) or a existing lattice is to be added to a file (`add_lattice` in `H5CUDS`).

Maybe to simplify your ctor of FileLattice you could just have:
```    def __init__(self, file, name=None, lattice=None):``` 
instead of
```    def __init__(self, file, name, type=None, base_vect=None, size=None, origin=None, record=None):```

Then like @itziakos  said, just check if `name` is not there and then you can query the `lattice` parameter to setup your lattice stored in the file.  If you do make such changes, I would also move the original 'updating' of nodes that you perform in `add_lattice` there.

**Comment by tuopuu** [comment]
@itziakos The FutureWarning is this: "FutureWarning: comparison to 'None' will result in an elementwise object comparison in the future." 
It refers to line 46 in file_lattice.py: "if None in [type, base_vect, size, origin]:"

**Comment by tuopuu** [comment]
@nathanfranklin, you're right, FileLattice constructor functionality overlaps with H5CUDS get_lattice/add_lattice in some ways. However, I think it's inevitable because we want to be able to create an in-file lattice without creating it fully in memory.

> Maybe to simplify your ctor of FileLattice you could just have:
> def __init__(self, file, name=None, lattice=None):

For lattice systems we currently always have a full numpy.ndarray (of at least empty elements, possibly with real CUBA-data) generated which allocates memory even if it is empty. Therefore first generating an empty Lattice that is used in construction of FileLattice is not very efficient, and in the case of big lattices, would possibly use more memory that is currently available. In the earlier stages of this PR I introduced a new class (BaseLattice) between ABCLattice and Lattice/FileLattice, that would have just the lattice attributes defined. With a BaseLattice using the constructor you suggested would work fine! However, it was suggested by @itziakos that using such in-between class should not be part of this PR, so we had to go back to passing the lattice attributes to FileLattice separately in order to have the functionality we need.

Just to summarize, we need from the constructor to
- give a reference to an existing lattice in H5CUDS-file if the name matches. In this case optional arguments can be undefined or given value None, because they are not used.
- create an empty lattice (with full pytables table) in H5CUDS-file if the name doesn't match, use optional parameters to describe it's shape, Record, etc. In this case, all optional arguments except the Record must be given, or it will result to an error message.
- not copy contents of an existing Lattice to H5CUDS-file: this is done by H5CUDS.add_lattice and FileLattice.update_node

I think a correct way to proceed is just to get rid of the FutureWarning message for now. I'll try to fix it somehow.

**Comment by itziakos** [comment]
> @itziakos The FutureWarning is this: "FutureWarning: comparison to 'None' will result in an elementwise object comparison in the future." 
> It refers to line 46 in file_lattice.py: "if None in [type, base_vect, size, origin]:"

Can you try using a tuple instead of a list?

**Comment by itziakos** [comment]
> I think a correct way to proceed is just to get rid of the FutureWarning message for now. I'll try to fix it somehow.

I agree with this statement.

**Comment by tuopuu** [comment]
> Can you try using a tuple instead of a list?

Unfortunately, using a tuple does not seem to help. I think comparing only the value of type is enough, because it's the first optional argument. I'll make a new commit shortly.

**Comment by itziakos** [comment]
> With a BaseLattice using the constructor you suggested would work fine! However, it was suggested by @itziakos that using such in-between class should not be part of this PR, so we had to go back to passing the lattice attributes to FileLattice separately in order to have the functionality we need.

Just to clarify, there are two problems that interact here:

- The current CUDS api does not allow creating new CUDS containers from the ABCEngine class. Touching this issue there are a number of discussion in the simphoni wiki and we should push for a more formal discussion/proposal in the wiki for the SSB to evalute.
-  The Lattice implementation preallocates space which, in combination with the current api limitation can lead to memory problems. There are a number of solutions for Lattice initialization issue, but I will suggest to wait since there are a number of changes in discussion for the api like `shared_data`, that will dictate better which solution is more appropriate.

Using a BaseLattice was not a solution, but a workaround which at the same time introduced other problems like overriding the ABCMeta restrictions of ABCLattice and introducing a partially implemented class.

**Comment by nathanfranklin** [comment]
> @nathanfranklin, you're right, FileLattice constructor functionality overlaps with H5CUDS get_lattice/add_lattice in some ways. However, I think it's inevitable because we want to be able to create an in-file lattice without creating it fully in memory.
>> Maybe to simplify your ctor of FileLattice you could just have:
>> def __init__(self, file, name=None, lattice=None):

> For lattice systems we currently always have a full numpy.ndarray (of at least empty elements, possibly with real CUBA-data) generated which allocates memory even if it is empty. **Therefore first generating an empty Lattice that is used in construction of FileLattice is not very efficient**, and in the case of big lattices, ...

*bold added by me

In my suggestion (``def __init__(self, file, name=None, lattice=None)`` in FileLattice), the only use of this __init__ of FileLattice where the `lattice` parameter would not equal `None` would be in the 'add_lattice' method in `H5CUDS`.  As this lattice already exists (as it was an argument in `add_lattice`, I don’t understand the concern about memory with my specific suggestion. 

I do see my suggestion forgot the ``record`` parameter.

Regardless, my suggestion can be ignored as its not important to closing this PR.

**Comment by kemattil** [comment]
Currently the standalone Lattice implementation indeed preallocates memory (numpy array of references to DataContainers, references initially = None). However, this is totally internal to the standalone Lattice implementation. If empty lattices are needed, you can change the implementation: postpone the memory allocation until first update to node data is executed.
Would this solve the problem?

**Comment by tuopuu** [comment]
> If empty lattices are needed, you can change the implementation: postpone the memory allocation until first update to node data is executed.

But this would require adding to lattice.update_node something like
if self._dc is not None:
   self._dc = np.empty(size, dtype=object)

which would be checked every time a node is updated. That could mean possibly billions of if checks for large lattices. One option would perhaps be to have two versions of update_node, one with the if check and one without, and use a reference/pointer to the correct function when calling update_node based on whether the array exist or not. Although I don't know if this is possible in Python.

**Comment by itziakos** [comment]
The PR looks ok to me, but we should to revisit the hdf5 backend and how containers are initialized.

**Comment by tuopuu** [comment]
So, this is OK to merge? We can create separate PR's that fix other issues, including the ones you mentioned.

**Comment by itziakos** [comment]
yup

**Comment by tuopuu** [comment]
Merged #74
